<div style="background-color:#9283BD;color:#ffffff;width:100%;display:inline-block;">
    <h1><strong>Aprendizaje Automático II</strong></h1>
</div>
<h2>Clase 9</h2>
<h3>Voting</h3>
<p>
El dataset incluye información de pacientes para la predicción de cancer.
</p>
<div style="background-color:#9283BD;color:#ffffff;width:100%;display:inline-block;">
    <b><h2>&nbsp Tabla de Contenidos</2></b>
</div>
<br>
<div>
    <ul>
        <li>
            <span>
                <a href="#intro">
                    <span>1</span>
                    <strong>Introducción</strong>
                </a>
            </span>
        </li>
        <li>
            <span>
                <a href="#importa">
                    <span>2</span>
                    <strong>Importación de librerías</strong>
                </a>
            </span>
        </li>
        <li>
            <span>
                <a href="#carga"><span>3</span>
                    <strong>Carga de datos</strong>
                </a>
            </span>
        </li>
        <li>
            <span>
                <a href="#division">
                    <span>4</span>
                    <strong>División en conjunto de entrenamiento y prueba</strong>
                </a>
            </span>
        </li>
        <li>
            <span>
                <a href="#modelos">
                    <span>5</span>
                    <strong>Creamos los diferentes modelos</strong>
                </a>
            </span>
        </li>
        <li>
            <span>
                <a href="#enseble">
                    <span>6</span>
                    <strong>Creamos el modelo de ensemble</strong>
                </a>
            </span>
        </li>
    </ul>
</div>

<a id="intro"></a>
<div style="background-color:#9283BD;color:#ffffff;width:100%;display:inline-block;">
    <b><h2>&nbsp 1. Introducción</h2></b>
</div>

Teoría
---
- Video de referencia "CIT IT22 - Machine Learning - Clase 9: Métodos de Ensamble": https://www.youtube.com/watch?v=GQ2Idkp0Syk

Práctica:
---
- Video "#MachineLearning Método de #Voting con #Python": https://www.youtube.com/watch?v=124yDoPFw2Y


<a id="importa"></a>
<div style="background-color:#9283BD;color:#ffffff;width:100%;display:inline-block;">
    <b><h2>&nbsp 2. Importación de librerías</h2></b>
</div>

In [2]:
import sklearn # Importar libreria de machine learning
from sklearn import model_selection # Importar libreria para partición de datos
from sklearn.linear_model import LogisticRegression # Regresión lineal
from sklearn.tree import DecisionTreeClassifier # Arbol de decisión
from sklearn.svm import SVC # Maquinas de soporte vectorial clasificación
from sklearn.ensemble import VotingClassifier # Modelo de votación (ensamble)

from sklearn.model_selection import train_test_split # Partición de datos

from sklearn import datasets # Importar datasets
import matplotlib
import matplotlib.pyplot as plt # Graficar

import warnings # Ignorar mensajes de advertencia
warnings.filterwarnings("ignore") # Ignorar mensajes de advertencia

import sys # Importar libreria para manejo de sistema y version de python

In [3]:
# Imprimimos las versiones de las librerias a utilizar
print('Versión de Python: {}'.format(sys.version))
print('Versión de scikit-learn: {}'.format(sklearn.__version__))
print('Versión de matplotlib: {}'.format(matplotlib.__version__))

# Particularmente use las siguientes versiones de librerias:
# Versión de Python: 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64 bit (AMD64)]
# Versión de scikit-learn: 1.3.1
# Versión de matplotlib: 3.7.2


Versión de Python: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]
Versión de scikit-learn: 1.5.2
Versión de matplotlib: 3.8.0


In [4]:
plt.style.use('ggplot') # Estilo de gráficos

<a id="carga"></a>
<div style="background-color:#9283BD;color:#ffffff;width:100%;display:inline-block;">
    <b><h2>&nbsp 3. Carga de datos</h2></b>
</div>

In [5]:
# Cargamos el dataset que nos provee la libreria sklearn
seed = 42 # Semilla para reproducibilidad
dataset = datasets.load_breast_cancer() # Cargamos el dataset

<a id="division"></a>
<div style="background-color:#9283BD;color:#ffffff;width:100%;display:inline-block;">
    <b><h2>&nbsp 4. División en conjunto de entrenamiento y prueba</h2></b>
</div>

In [6]:
X = dataset.data # variables independientes features
y = dataset.target # variable dependiente target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25) # Partición de datos
kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True) # Validación cruzada

<a id="modelos"></a>
<div style="background-color:#9283BD;color:#ffffff;width:100%;display:inline-block;">
    <b><h2>&nbsp 5. Creamos los diferentes modelos</h2></b>
</div>

In [7]:
estimators = [] # Lista de modelos a utilizar
model1 = LogisticRegression() # Modelo de regresión logistica
estimators.append(('logistic', model1)) # Agregamos el modelo a la lista

model2 = DecisionTreeClassifier() # Modelo de arbol de decisión
estimators.append(('tree', model2)) # Agregamos el modelo a la lista

model3 = SVC() # Modelo de maquinas de soporte vectorial
estimators.append(('svm', model3)) # Agregamos el modelo a la lista

<a id="enseble"></a>
<div style="background-color:#9283BD;color:#ffffff;width:100%;display:inline-block;">
    <b><h2>&nbsp 6. Creamos el modelo de ensemble</h2></b>
</div>

In [8]:
ensemble = VotingClassifier(estimators) # Modelo de votación (ensamble)
results = model_selection.cross_val_score(ensemble, X_train, y_train, cv=kfold) # Validación cruzada
print() # Imprimimos un salto de linea
print(results.mean()) # Imprimimos el promedio de los resultados


0.9412513842746402




------------------

# Analizaremos el dataset wine con SVM

In [30]:
# Importando las librerías necesarias
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.ensemble import VotingClassifier
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.metrics import accuracy_score, classification_report

In [31]:
# Cargando el dataset "wine"
wine_data = load_wine()
X, y = wine_data.data, wine_data.target

In [32]:
# Dividiendo el dataset en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [33]:
# Creando un modelo SVM con un pipeline de preprocesamiento (escalado de datos)
svm_model = make_pipeline(StandardScaler(), SVC(probability=True, random_state=42))

In [34]:
# Creando el clasificador de voting con un solo modelo (SVM)
voting_clf = VotingClassifier(estimators=[('svm', svm_model)], voting='soft')

In [35]:
# Entrenando el modelo
voting_clf.fit(X_train, y_train)

VotingClassifier(estimators=[('svm',
                              Pipeline(steps=[('standardscaler',
                                               StandardScaler()),
                                              ('svc',
                                               SVC(probability=True,
                                                   random_state=42))]))],
                 voting='soft')

In [36]:
# Realizando predicciones
y_pred = voting_clf.predict(X_test)

In [40]:
# Evaluación del modelo
precision = accuracy_score(y_test, y_pred)
reporte = classification_report(y_test, y_pred)

print(precision)
print(reporte)

0.9722222222222222
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       0.93      1.00      0.97        14
           2       1.00      0.90      0.95        10

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.97        36
weighted avg       0.97      0.97      0.97        36



# Ahora realizaremos el procedimiento con KNN

In [41]:
from sklearn.neighbors import KNeighborsClassifier

# Creando un modelo KNN
knn_model = KNeighborsClassifier()

In [42]:
# Creando el clasificador de voting con un solo modelo (KNN)
voting_clf_knn = VotingClassifier(estimators=[('knn', knn_model)], voting='soft')

In [43]:
# Entrenando el modelo
voting_clf_knn.fit(X_train, y_train)

VotingClassifier(estimators=[('knn', KNeighborsClassifier())], voting='soft')

In [44]:
# Realizando predicciones
y_pred_knn = voting_clf_knn.predict(X_test)

In [45]:
# Evaluación del modelo
accuracy_knn = accuracy_score(y_test, y_pred_knn)
report_knn = classification_report(y_test, y_pred_knn)

In [47]:
print(accuracy_knn)
print(report_knn)

0.8055555555555556
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       0.77      0.71      0.74        14
           2       0.64      0.70      0.67        10

    accuracy                           0.81        36
   macro avg       0.80      0.80      0.80        36
weighted avg       0.81      0.81      0.81        36



# Por último realizaremos el procedimiento con árbol de decisión

In [49]:
from sklearn.tree import DecisionTreeClassifier

# Creando un modelo de Árbol de Decisión
dt_model = DecisionTreeClassifier(random_state=42)

In [50]:
# Creando el clasificador de voting con un solo modelo (Árbol de Decisión)
voting_clf_dt = VotingClassifier(estimators=[('dt', dt_model)], voting='soft')

In [51]:
# Entrenando el modelo
voting_clf_dt.fit(X_train, y_train)

VotingClassifier(estimators=[('dt', DecisionTreeClassifier(random_state=42))],
                 voting='soft')

In [52]:
# Realizando predicciones
y_pred_dt = voting_clf_dt.predict(X_test)

In [53]:
# Evaluación del modelo
accuracy_dt = accuracy_score(y_test, y_pred_dt)
report_dt = classification_report(y_test, y_pred_dt)

print(accuracy_dt)
print(report_dt)

0.9444444444444444
              precision    recall  f1-score   support

           0       1.00      0.92      0.96        12
           1       0.88      1.00      0.93        14
           2       1.00      0.90      0.95        10

    accuracy                           0.94        36
   macro avg       0.96      0.94      0.95        36
weighted avg       0.95      0.94      0.94        36



# Análisis y documentación de los resultados
Modelos probados

Se utilizaron tres algoritmos de clasificación como base del esquema de ensamble con el método de voting. Cada modelo fue evaluado de manera independiente con los siguientes resultados:

Support Vector Machine (SVM):

Precisión general: 97.22%

Desempeño sobresaliente, especialmente con la clase mayoritaria (clase 0), logrando una precisión perfecta.
Su desempeño sugiere que es efectivo para datos bien separados en el espacio de características.

K-Nearest Neighbors (KNN):

Precisión general: 80.56%

El desempeño más bajo entre los tres clasificadores. Aunque obtuvo buenos resultados en la clase 0, su capacidad para clasificar correctamente las clases 1 y 2 fue limitada.
Esto podría deberse a la sensibilidad de KNN a la escala de los datos y al ruido en el dataset.

Árbol de Decisión:

Precisión general: 94.44%

Muy buen desempeño global, aunque ligeramente inferior al modelo SVM.
Los resultados muestran que es capaz de manejar bien los datos y distinguir las tres clases, con menor precisión en la clase 1.
Conclusiones
SVM fue el modelo con mejor desempeño en términos de precisión general y métricas por clase. Esto es consistente con su capacidad para encontrar un hiperplano óptimo en datos con separación clara.
KNN no logró un rendimiento competitivo en este conjunto de datos, lo que indica que podría no ser el mejor candidato para el problema, especialmente sin ajustar parámetros como el número de vecinos o las métricas de distancia.
Árbol de Decisión ofrece una solución interpretativa y robusta con buen equilibrio entre precisión y recall, pero no alcanza el desempeño de SVM en este caso.


# Posibles mejoras

Combinar estos clasificadores en un esquema de ensamble de voting que aproveche las fortalezas de cada modelo podría mejorar aún más el rendimiento.
Realizar ajuste de hiperparámetros para los modelos de base, por ejemplo:

Cambiar el kernel o el parámetro C en el modelo SVM.

Ajustar el número de vecinos y las métricas en KNN.

Ajustar la profundidad maxima del arbol de decision.

Evaluar el impacto de otras técnicas de preprocesamiento, como la selección de características o la normalización avanzada.

# Utilizando el mismo dataset "wine", implementaremos un modelo de clasificación empleando el método de ensamble bagging, basado en SVM

In [55]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.ensemble import BaggingClassifier
from sklearn.metrics import accuracy_score, classification_report

# Cargar el dataset "wine"
wine_data = load_wine()
X, y = wine_data.data, wine_data.target

In [56]:
# Dividir los datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [59]:
# Creando el modelo Bagging con base SVM
bagging_svm = BaggingClassifier(
    estimator=SVC(probability=True, random_state=42),
    n_estimators=10,
    random_state=42
)

In [60]:
# Entrenar el modelo
bagging_svm.fit(X_train, y_train)


BaggingClassifier(estimator=SVC(probability=True, random_state=42),
                  random_state=42)

In [61]:
# Realizar predicciones
y_pred = bagging_svm.predict(X_test)

In [62]:
# Evaluar el modelo
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

In [63]:
# Mostrar resultados
print(f"Precisión general: {accuracy:.2f}")
print("Reporte de clasificación:")
print(report)

Precisión general: 0.69
Reporte de clasificación:
              precision    recall  f1-score   support

           0       0.92      0.92      0.92        12
           1       0.58      1.00      0.74        14
           2       0.00      0.00      0.00        10

    accuracy                           0.69        36
   macro avg       0.50      0.64      0.55        36
weighted avg       0.53      0.69      0.59        36



Comparación entre Voting y Bagging
Rendimiento de los métodos

Voting (con SVM):

Precisión general: 97.22%

Reporte de clasificación:

Clase 0: Excelente desempeño con precisión y recall perfectos.
Clase 1 y 2: Resultados consistentes, con f1-scores altos.

Bagging (con SVM):

Precisión general: 69%

Reporte de clasificación:

Clase 0: Buen desempeño (precisión y recall de 92%).

Clase 1: Alto recall (100%), pero baja precisión (58%).

Clase 2: Desempeño deficiente (f1-score de 0.00).


Análisis de las diferencias observadas

Precisión general:

Voting supera ampliamente a Bagging en este dataset. Esto se debe a que el modelo Voting combina predicciones de varios modelos fuertes, mientras que Bagging se beneficia más de clasificadores débiles, como Árboles de Decisión.

Desempeño por clase:

Voting distribuye mejor su capacidad predictiva entre todas las clases, mostrando buenos resultados para las clases minoritarias (1 y 2).
Bagging, en cambio, lucha para manejar la clase 2 (minoritaria), ya que las instancias de esta clase son menos representadas en los subconjuntos de datos utilizados durante el ensamble.

Naturaleza del método:

Voting combina modelos independientes que pueden tener arquitecturas o características diferentes, aprovechando la diversidad en los clasificadores para mejorar la precisión.
Bagging es más adecuado cuando se busca reducir la varianza y prevenir sobreajuste, siendo más efectivo con clasificadores altamente inestables (como Árboles de Decisión).

Casos en los que uno podría ser más adecuado que el otro

Voting:

Ideal para datasets con clases bien balanceadas o con suficiente separación en las características.
Beneficioso cuando se dispone de múltiples clasificadores con buen rendimiento individual.
Preferible cuando se requiere precisión en todas las clases, ya que Voting tiende a aprovechar las fortalezas de cada modelo.

Bagging:

Funciona mejor con clasificadores débiles o inestables, como árboles de decisión o regresores lineales en datos no lineales.

Más adecuado para datasets con ruido o cuando hay alto riesgo de sobreajuste en modelos individuales.

Menos sensible a los datos desbalanceados, pero necesita ajustes adicionales en el tamaño de las submuestras.

Conclusión

En este caso específico, el método Voting mostró un desempeño superior al método Bagging al trabajar con el clasificador SVM. Esto se debe a que SVM es un modelo estable que no se beneficia significativamente del ensamble tipo Bagging. Sin embargo, Bagging podría mejorar si se usa con clasificadores como Árboles de Decisión o si se ajustan hiperparámetros como el tamaño de las submuestras.

In [64]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.metrics import accuracy_score, classification_report

# Cargar el dataset "wine"
wine_data = load_wine()
X, y = wine_data.data, wine_data.target

In [66]:
# Dividir los datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Crear el modelo Bagging con Árboles de Decisión como base
bagging_dt = BaggingClassifier(
    estimator=DecisionTreeClassifier(random_state=42),  # Changed 'base_estimator' to 'estimator'
    n_estimators=10,  # Número de clasificadores base
    random_state=42
)

In [67]:
# Entrenar el modelo
bagging_dt.fit(X_train, y_train)

# Realizar predicciones
y_pred = bagging_dt.predict(X_test)

In [69]:
# Evaluar el modelo
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

# Mostrar resultados
print(f"Precisión general: {accuracy:.2f}")
print("Reporte de clasificación:")
print(report)

Precisión general: 1.00
Reporte de clasificación:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       1.00      1.00      1.00        14
           2       1.00      1.00      1.00        10

    accuracy                           1.00        36
   macro avg       1.00      1.00      1.00        36
weighted avg       1.00      1.00      1.00        36



Resultados obtenidos
Voting (con SVM):


Precisión general: 97.22%

Reporte de clasificación:

Excelente desempeño en todas las clases, pero con ligeras caídas en las métricas para la clase 1 y la clase 2.
Menor sensibilidad en comparación con Bagging en el manejo de clases minoritarias.

Bagging (con Árboles de Decisión):

Precisión general: 100%

Reporte de clasificación:

Desempeño perfecto en todas las métricas y clases. Cada clase fue clasificada correctamente, logrando precisión, recall y f1-score perfectos.
Análisis de las diferencias observadas

Precisión general:

Bagging con Árboles de Decisión logró una precisión perfecta, mientras que Voting (con SVM) obtuvo un 97.22%.
Esto indica que el ensamble Bagging, al utilizar clasificadores inestables como los árboles de decisión, puede manejar mejor las particularidades del dataset "wine".

Desempeño por clase:

En Bagging, todas las clases fueron clasificadas perfectamente, destacando su capacidad para manejar tanto clases mayoritarias como minoritarias.
Voting mostró un muy buen desempeño, pero podría verse limitado al trabajar con un único modelo base (SVM), que depende del espacio de características y la separación entre las clases.
Robustez del modelo:

Bagging reduce la varianza del modelo base (Árboles de Decisión), aumentando la robustez frente al sobreajuste y logrando un mejor rendimiento en este caso.
Voting se beneficia al combinar la capacidad de diferentes clasificadores, pero cuando se basa únicamente en un modelo, su capacidad está limitada por la estabilidad y el rendimiento del clasificador elegido.
Casos en los que uno podría ser más adecuado que el otro

Voting:

Es más adecuado cuando:
Se cuenta con clasificadores fuertes y diversos.
El dataset tiene una separación clara entre clases.
Se requiere un balance entre interpretabilidad y rendimiento.

Bagging:

Es más adecuado cuando:

Se trabaja con clasificadores inestables (como Árboles de Decisión).
El dataset es propenso al sobreajuste.
Hay ruido en los datos o la separación entre clases no es tan evidente.

Conclusión
En este caso, el método Bagging basado en Árboles de Decisión mostró un rendimiento superior, logrando una precisión perfecta en todas las clases. Esto resalta su capacidad para aprovechar clasificadores inestables y generar predicciones más robustas en datasets con características similares al "wine".

### Para mejorar el rendimiento del esquema de Voting y obtener una mayor robustez, podemos combinar diferentes clasificadores en lugar de usar solo uno (como SVM). Al integrar varios modelos con diferentes características y fortalezas, podemos aprovechar mejor la diversidad del ensemble.

### A continuación te muestro cómo implementar un esquema de Voting con tres clasificadores: SVM, Árbol de Decisión y K-Nearest Neighbors (KNN). Este enfoque permite combinar las fortalezas de cada uno y mejorar la precisión general del modelo.

In [71]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.metrics import accuracy_score, classification_report

In [72]:
# Cargar el dataset "wine"
wine_data = load_wine()
X, y = wine_data.data, wine_data.target

In [73]:
# Dividir los datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [74]:
# Definir los clasificadores base
svm_model = SVC(probability=True, random_state=42)
dt_model = DecisionTreeClassifier(random_state=42)
knn_model = KNeighborsClassifier()

In [75]:
# Crear el clasificador de Voting con los tres clasificadores base
voting_clf = VotingClassifier(
    estimators=[('svm', svm_model), ('dt', dt_model), ('knn', knn_model)],
    voting='soft'  # 'soft' para usar probabilidades en lugar de predicciones duras
)


In [76]:
# Entrenar el modelo
voting_clf.fit(X_train, y_train)

VotingClassifier(estimators=[('svm', SVC(probability=True, random_state=42)),
                             ('dt', DecisionTreeClassifier(random_state=42)),
                             ('knn', KNeighborsClassifier())],
                 voting='soft')

In [77]:
# Realizar predicciones
y_pred = voting_clf.predict(X_test)

In [78]:
# Evaluar el modelo
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

In [79]:
# Mostrar resultados
print(f"Precisión general: {accuracy:.2f}")
print("Reporte de clasificación:")
print(report)

Precisión general: 0.97
Reporte de clasificación:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        12
           1       0.93      1.00      0.97        14
           2       1.00      0.90      0.95        10

    accuracy                           0.97        36
   macro avg       0.98      0.97      0.97        36
weighted avg       0.97      0.97      0.97        36



Resultados obtenidos

Bagging con Árbol de Decisión:

Precisión general: 100%

Reporte de clasificación:

Excelente desempeño, con precisión, recall y f1-score perfectos para todas las clases.

Voting con SVM, Árbol de Decisión y KNN:

Precisión general: 97%

Reporte de clasificación:

Clase 0: Precisión y recall perfectos (1.00).

Clase 1: Alta precisión (93%) y recall perfecto (100%).

Clase 2: Buen desempeño (precisión y recall cerca de 1.00).

En general, buen desempeño con ligeras caídas en la clase 1.

Diferencias observadas

Precisión general:

Bagging con Árbol de Decisión alcanzó una precisión perfecta del 100%, mientras que el esquema de Voting con 3 clasificadores alcanzó una precisión de 97%. Esto sugiere que Bagging con un clasificador inestable como el Árbol de Decisión es muy efectivo al reducir la varianza, generando un modelo robusto que maneja bien todas las clases del dataset.

Desempeño por clase:

Bagging con Árbol de Decisión mostró un desempeño perfecto en todas las clases, incluyendo la clase 2 (que tiene menos ejemplos).

Voting con SVM, Árbol de Decisión y KNN también mostró buen desempeño, con clases 0 y 2 alcanzando una precisión perfecta. Sin embargo, para la clase 1, el KNN parece ser menos efectivo, lo que afectó ligeramente la precisión para esa clase (93% de precisión frente al 100% del Bagging).

Robustez y comportamiento en clases minoritarias:

Bagging manejó mejor las clases minoritarias (como la clase 2) al combinar múltiples árboles de decisión, lo que permitió reducir la varianza y lograr una predicción consistente.

Voting, al combinar modelos como SVM, Árbol de Decisión y KNN, no pudo igualar el rendimiento de Bagging en cuanto a la consistencia de las predicciones, ya que los modelos base tienen fortalezas y debilidades distintas, lo que puede afectar el rendimiento en clases minoritarias.

Variabilidad en el rendimiento:

Bagging ofreció una precisión más estable, sin caídas notables en ninguna de las clases.

Voting mostró ligeras variaciones en la clase 1, especialmente debido a la diferencia en cómo KNN maneja las distancias en los puntos cercanos.
Análisis de casos adecuados para cada enfoque

Bagging con Árbol de Decisión:

Ideal para datasets con ruido o cuando se necesita reducir el sobreajuste. Bagging es un enfoque efectivo cuando se trabaja con clasificadores inestables, como Árboles de Decisión, ya que ayuda a estabilizar las predicciones.
Cuando las clases son desbalanceadas o tienen una variabilidad significativa, Bagging es muy útil porque reduce la varianza y mejora la precisión general.
Casos de uso: Cuando los datos son ruidosos o contienen patrones complejos que pueden hacer que los modelos simples se sobreajusten, Bagging ofrece una gran robustez.

Voting con 3 clasificadores (SVM, Árbol de Decisión y KNN):

Ideal cuando se tiene un conjunto diverso de clasificadores fuertes y se desea aprovechar sus diferentes características. Por ejemplo, SVM es eficaz en espacios de alta dimensión, Árboles de Decisión manejan bien datos no lineales y KNN es sensible a la proximidad local de los datos.

Cuando se tiene confianza en la variedad de los modelos base: Voting es particularmente útil cuando los clasificadores son lo suficientemente buenos por sí solos, pero se puede aprovechar su combinación para mejorar el rendimiento general.

Casos de uso: Cuando el conjunto de datos no es ruidoso y tiene una estructura relativamente clara, y se busca mejorar la precisión utilizando la diversidad de modelos. Además, el enfoque 'soft' voting puede ser beneficioso cuando se desea aprovechar las probabilidades de los clasificadores para obtener un voto más equilibrado.

Conclusión

Bagging con Árbol de Decisión ofrece un rendimiento excelente y más estable en este caso, especialmente para manejar clases minoritarias y reducir la varianza. Es el método más adecuado si se busca precisión perfecta y robustez frente al ruido.

Voting con 3 clasificadores proporciona un buen rendimiento general, especialmente cuando se tienen clasificadores con características complementarias. Sin embargo, su rendimiento podría verse afectado por la variabilidad entre los clasificadores, especialmente si uno de ellos no se desempeña bien con ciertas clases.

En resumen, si el objetivo es maximizar la precisión y robustez, Bagging con Árboles de Decisión es el mejor enfoque. Si se dispone de clasificadores fuertes y diversos y se busca flexibilidad, Voting podría ser una opción adecuada.